internal/pkg/build/buildkit/daemon/daemon.go

```diff
@@ -0,0 +1,695 @@
+	if err != nil {
+		return nil, err
+	}
+	opt.GCPolicy = getGCPolicy(cfg.GCConfig, common.config.Root)
+	opt.BuildkitVersion = getBuildkitVersion()
+	opt.RegistryHosts = hosts
+
+	if platformsStr := cfg.Platforms; len(platformsStr) != 0 {
+		platforms, err := parsePlatforms(platformsStr)
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid platforms")
+		}
+		opt.Platforms = platforms
+	}
+	w, err := base.NewWorker(ctx, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	return []worker.Worker{w}, nil
+}
+
+func snapshotterFactory(_ context.Context, commonRoot string, cfg config.OCIConfig) (BkSnapshotterFactory, error) {
+	var (
+		name    = cfg.Snapshotter
+		address = cfg.ProxySnapshotterPath
+	)
+	if address != "" {
+		snFactory := BkSnapshotterFactory{
+			Name: name,
+		}
+		if _, err := os.Stat(address); os.IsNotExist(err) {
+			return snFactory, errors.Wrapf(err, "snapshotter doesn't exist on %q (Do not include 'unix://' prefix)", address)
+		}
+		snFactory.New = func(root string) (ctdsnapshot.Snapshotter, error) {
+			backoffConfig := backoff.DefaultConfig
+			backoffConfig.MaxDelay = 3 * time.Second
+			connParams := grpc.ConnectParams{
+				Backoff: backoffConfig,
+			}
+			gopts := []grpc.DialOption{
+				grpc.WithTransportCredentials(insecure.NewCredentials()),
+				grpc.WithConnectParams(connParams),
+				grpc.WithContextDialer(dialer.ContextDialer),
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(defaults.DefaultMaxRecvMsgSize)),
+				grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(defaults.DefaultMaxSendMsgSize)),
+			}
+			conn, err := grpc.Dial(dialer.DialAddress(address), gopts...)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to dial %q", address)
+			}
+			return snproxy.NewSnapshotter(snapshotsapi.NewSnapshotsClient(conn), name), nil
+		}
+		return snFactory, nil
+	}
+
+	if name == "auto" {
+		if err := overlayutils.Supported(commonRoot); err == nil {
+			name = "overlayfs"
+		} else {
+			sylog.Debugf("auto snapshotter: overlayfs is not available for %s, trying fuse-overlayfs: %v", commonRoot, err)
+			if err2 := fuseoverlayfs.Supported(commonRoot); err2 == nil {
+				name = "fuse-overlayfs"
+			} else {
+				sylog.Debugf("auto snapshotter: fuse-overlayfs is not available for %s, falling back to native: %v", commonRoot, err2)
+				name = "native"
+			}
+		}
+		sylog.Infof("auto snapshotter: using %s", name)
+	}
+
+	snFactory := BkSnapshotterFactory{
+		Name: name,
+	}
+	switch name {
+	case "overlayfs": // not "overlay", for consistency with containerd snapshotter plugin ID.
+		snFactory.New = func(root string) (ctdsnapshot.Snapshotter, error) {
+			return overlay.NewSnapshotter(root, overlay.AsynchronousRemove)
+		}
+	default:
+		return snFactory, errors.Errorf("unknown snapshotter name: %q", name)
+	}
+	return snFactory, nil
+}
+
+func validOCIBinary() bool {
```

Comment:
Elsewhere we set the runtime to possibly be `crun`, but this is inferring that either `runc` or `buildkit-runc` still need to be present or the worker will error out?

Reply:
This function (`validOCIBinary()`) was only getting called under a very particular circumstance, where the overall `cfg.Workers.OCI.Enabled` was `nil` (yes it's a pointer... don't ask) - which is a circumstance that doesn't actually arise because of hard-coded changes we made relative to the original moby/buildkit code. I'll just remove it completely - if there's no `runc` _or_ `crun` installed, it'll error out in the call to Runtime() that occurs in `ociWorkerInitializer()`.